### DISCUSSION TITLE
Generating *.Designer.cs from .resx

### DISCUSSION THREAD

**Discussion by Xwilarg** [community/community#26244] [Actions]
<p>Hi,<br>
I’m using C# on a project and I have resx files that generate some Designer.cs files, however since they are generated I would like to not push them on Git.<br>
However the Build phase of GitHub action doesn’t generate them, but the Test phase do<br>
I’m building with</p>
<pre><code class="lang-auto">dotnet build --configuration Release --no-restore
</code></pre>
<p>And testing with</p>
<pre><code class="lang-auto">dotnet test --no-restore --verbosity normal --collect:"XPlat Code Coverage" --settings coverlet.runsettings
</code></pre>
<p>I noticed that resgen.exe is used to generate them but I wasn’t able to call it.</p>
<p>So my question is, would someone know how I can generate Designer files from resx in GitHub Actions?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/Xwilarg">@Xwilarg</a>,</p>
<p>I’m afraid you can’t generate a <strong>.Designer.cs</strong> from a <strong>.resx</strong> in your WinForm application.<br>
To understand this point, you need to understand what is resource file and what is designer file at first:</p>
<ol>
<li>
<p>The <strong>.resx</strong> is one of the resource file types. The resource file types include text-based resource file (<strong>.txt</strong> and <strong>.restext</strong>), XML-based resource file (<strong>.resx</strong>).<br>
A resource is any nonexecutable data that is logically deployed with an app. Resources can contain data in a number of forms, including strings, images, and persisted objects. Storing your data in a resource file enables you to change the data without recompiling your entire app. It also enables you to store data in a single location, and eliminates the need to rely on hard-coded data that is stored in multiple locations. More deatils, see “<a href="https://docs.microsoft.com/en-us/dotnet/framework/resources/" rel="nofollow noopener"><strong>Resources in .NET apps</strong></a>”.</p>
</li>
<li>
<p>The designer file (<strong>.Designer.cs</strong>) is a code file automatically generated by your designer to hold the form’s layout information that was created using the Visual Studio IDE. Once you add a new form in your application, VS will automatically generate the designer file for this form.</p>
</li>
</ol>
<p>According to the above introductions, logically, there is no relationship between the resource file and the designer file. The designer file (<strong>.Designer.cs</strong>) is required for the WinForm application, you can’t ignore it in your project.</p>
<blockquote>
<p>I noticed that <strong>resgen.exe</strong> is used to generate them but I wasn’t able to call it.</p>
</blockquote>
<p>The Resource File Generator (<strong>Resgen.exe</strong>) is used to convert the resource files to common language runtime binary (.resources) files that can be embedded in a runtime binary executable or satellite assembly.<br>
This tool is automatically installed with Visual Studio. To run the tool, use the Developer Command Prompt for Visual Studio (or the Visual Studio Command Prompt in Windows 7).<br>
It can’t be used to generate the designer files from a the resource files. More details, see “<a href="https://docs.microsoft.com/en-us/dotnet/framework/tools/resgen-exe-resource-file-generator" rel="nofollow noopener"><strong>Resgen.exe (Resource File Generator)</strong></a>”.</p>

**Comment by Xwilarg** [comment]
<p>Thanks for your answer,<br>
My application is a console application and not a WinForm one (but I’m not sure if that really matters)<br>
I’m using it because it’s easier and faster for me than creating a dictionary or a JSON file.<br>
Visual Studio already generate the Designer file from the resx file thanks to this code in the csproj</p>
<pre><code class="lang-auto">  &lt;ItemGroup&gt;
    &lt;Compile Update="LanguageResources\Hiragana.Designer.cs"&gt;
      &lt;DesignTime&gt;True&lt;/DesignTime&gt;
      &lt;AutoGen&gt;True&lt;/AutoGen&gt;
      &lt;DependentUpon&gt;Hiragana.resx&lt;/DependentUpon&gt;
    &lt;/Compile&gt;
  &lt;/ItemGroup&gt;

  &lt;ItemGroup&gt;
    &lt;EmbeddedResource Update="LanguageResources\Hiragana.resx"&gt;
      &lt;Generator&gt;ResXFileCodeGenerator&lt;/Generator&gt;
      &lt;LastGenOutput&gt;Hiragana.Designer.cs&lt;/LastGenOutput&gt;
    &lt;/EmbeddedResource&gt;
  &lt;/ItemGroup&gt;
</code></pre>
<p>As you said, it’s generated by VS and I would have like to force the generation by GitHub CI<br>
But maybe I’m not doing things the right way?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/Xwilarg">@Xwilarg</a>,</p>
<p>If you can use the  <strong>.resx</strong> file to generate the <strong>.Designer.cs</strong> file on your local environment, maybe you can try to list the completed steps you have used, and check the command lines you were using to execute these steps. Then you can try to run these command lines in your workflow to see if it can work.</p>

**Comment by Xwilarg** [comment]
<p>The issue is that I’m not using any command, they are automatically generated.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/Xwilarg">@Xwilarg</a>,</p>
<p>Of course, as mentioned above, it is generally the <strong>.Designer.cs</strong> file is automatically generated by VS, it is not generated from the <strong>.resx</strong> file. You may have some misunderstandings about the <strong>.resx</strong> file.</p>

**Comment by Xwilarg** [comment]
<p>Yeah I’m indeed a bit confused by that, since the .Designer.cs contains what is inside the .resx and a way to access it in C#, why isn’t it “generated from it”?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/Xwilarg">@Xwilarg</a>,</p>
<p>Sometimes, the <strong>.Designer.cs</strong> may reference some strings defined in the resource file to set the contents displayed on the application UI. But this does not mean the <strong>.Designer.cs</strong> is generated from the resource file.<br>
Other files in the project may also reference the data defined in the resource file, that also does not mean these files are generated from the resource file.</p>
<p>In the <strong>.csproj</strong> file, normally it will list all the required source files of the application.</p>
<ul>
<li>
<p>The <strong>.Designer.cs</strong> defines the UI of the application, it is required. Maybe, sometimes the <strong>.Designer.cs</strong> defines an empty or blank UI when the application does not need a UI.</p>
</li>
<li>
<p>The <strong>.resx</strong> file defines some data may be referenced by other source files in the same project.</p>
</li>
</ul>
<p>Essentially, there is no relationship between the <strong>.Designer.cs</strong> and the <strong>.resx</strong>.<br>
The <strong>.Designer.cs</strong> can reference the data defined in the <strong>.resx</strong>, or no need to reference <strong>.resx</strong>. you can directly define all the needed UI strings in the <strong>.Designer.cs</strong>.</p>

**Comment by z9security** [comment]
<p>Inside my *Designer.cs file:</p>
<pre><code class="lang-auto">    // This class was auto-generated by the StronglyTypedResourceBuilder
    // class via a tool like ResGen or Visual Studio.
    // To add or remove a member, edit your .ResX file then rerun ResGen
</code></pre>
<p>I think the comments downplaying the connection between the 2 files can’t be applicable in all cases.  In my case this file is absolutely fully generated from the .ResX file.</p>
<p>I was able to fully generate the file .Designer.cs file using the information here:</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <a href="https://docs.microsoft.com/en-us/dotnet/framework/tools/resgen-exe-resource-file-generator" target="_blank" rel="noopener nofollow ugc">docs.microsoft.com</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/45845681/181096406-e14e51b3-ad24-40af-833e-aae50cfc61c2.png" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://docs.microsoft.com/en-us/dotnet/framework/tools/resgen-exe-resource-file-generator" target="_blank" rel="noopener nofollow ugc">Resgen.exe (Resource File Generator)</a></h3>

<p>Use Resgen.exe, the Resource File Generator. Convert text (.txt, .restext) and XML resource format (.resx) files to embeddable CLR runtime binaries (.resources).</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>using the /str option.</p>

**Comment by Entiat** [comment]
<p>Can you elaborate a bit more on the command line to resgen.exe that can regenerate .designer.cs files? For example, sample command lines?</p>
<p>I’ve poked at every command line option in the help, and can’t find the magic incantation that does anything other than generate resources.resources.  I’m sure I’ll smack my forehead and say, “doh” when I see it…but I’m having a heck of a time doing it now (Visual Studio steadfastly refuses to regenerate the .designer.cs file, no matter what I poke in the project/solution).</p>
<p>Thanks.</p>

**Reply by DanielHolland** [reply]
This is probably years too late but hopefully it helps someone else on this search, courtesy of ChatGPT:
```
cd "C:\Projects\MyApp\Resources"
resgen.exe MyAppResources.resx `
   /str:cs,MyApp.Resources,MyAppResources,MyAppResources.Designer.cs `
   /publicClass
```

This will generate both a binary and designer file for your "MyAppResources.resx" resources file

**Comment by Luiz-Monad** [comment]
I had this problem, I found a solution by reading the `powershell` source code.    [PowerShell://docs/building/internals.md](https://github.com/PowerShell/PowerShell/blob/7dc4587014bfa22919c933607bf564f0ba53db2e/docs/building/internals.md#resgen)
The idea is simple, you compile a helper assembly that generates `resx` then you use a Source Generator to generate `cs` from that.
https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/

references: https://github.com/dotnet/sdk/issues/94#issuecomment-1333163978
